src/lib/data/bluetooth.ts

```diff
@@ -0,0 +1,168 @@
+import {Sensor} from '../../components/Sensor';
+import {RealtimeDataModel, SensorModel, db} from '../../database';
+import {
+  cadenceMeter,
+  getAllSensors,
+  heartRateMonitor,
+  powerMeter,
+} from '../sensor';
+import {getOrCreateRealtimeRecord} from './realtime';
+import {Subscription} from 'rxjs';
+
+type Sensor = {
+  id: string;
+  name: string;
+  is_primary: boolean;
+  sensorType: string[];
+  type: string;
+  address: string;
+};
+
+function getSensorFromType(sensorType: string) {
+  switch (sensorType) {
+    case 'HeartRate':
+      return heartRateMonitor;
+
+    case 'CyclingPower':
+      return powerMeter;
+
+    case 'CyclingSpeedAndCadence':
+      return cadenceMeter;
+
+    default:
+      console.log('Unknown sensor type');
+      throw new Error('Unknown sensor type');
+  }
+}
+
+// Find first sensor of specific type in DB. Can't query it because table is JSON :/
+function findFirstSensorOfType(
+  sensors: Sensor[],
+  sensorType: string,
+): Sensor | undefined {
+  return sensors.find(sensor => sensor.sensorType.includes(sensorType));
+}
+
+export async function onHeartRateSensorEvent(
+  data: any,
+): Promise<RealtimeDataModel> {
+  // When we get new Power data write it to realtime table
+  let record = await getOrCreateRealtimeRecord();
+  if (record.heartRate === data.bpm) {
+    // Only write to realtime table if we have a difference in value
+    return record;
+  }
+  return db.write(async () => {
+    return record.update(() => {
+      record.heartRate = data.bpm;
+      return record;
+    });
+  });
+}
+
+export async function onPowerSensorEvent(
+  data: any,
+): Promise<RealtimeDataModel> {
+  // When we get new HR data write it to realtime table
+  console.log(data);
+  let record = await getOrCreateRealtimeRecord();
+  if (record.instantPower === data.instantaneous_power) {
+    // Only write to realtime table if we have a difference in value
+    return record;
+  }
+  return db.write(async () => {
+    return record.update(() => {
+      record.instantPower = data.instantaneous_power;
+      return record;
+    });
+  });
+}
+
+async function connectToSensor(
+  sensor: Sensor | undefined,
+  bleSensor: any,
+  sensorType: string,
+  callback: {
+    (r: RealtimeDataModel): Promise<RealtimeDataModel>;
+    (r: RealtimeDataModel): Promise<RealtimeDataModel>;
+  },
+) {
+  bleSensor = getSensorFromType(sensorType);
+  if (sensor === undefined) {
+    throw new Error(`No ${sensorType} sensor found`);
+  }
+
+  try {
+    // Try connect to our BLE sensor
+    bleSensor.address = sensor.address;
```

Comment:
Shouldn't this be await sensor.connect()? Assigning it to the global Ble sensor seems very bug prone as you'd need to co-ordinate between other sensor types right?

Reply:
hmm, yeah this doesn't look correct, but the bleSensor here would actually be set to heartRateMonitor global, which is correct. But i think I need to rework this because the naming is terrible here, having sensor and bleSensor is just super confusing :/